hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+# Versioning
+
+Specification for how storage versioning will operate for code generated CRD definitions.
+
+## Goals
+
+**Auto-generated conversions:** As far as practical, we want to autogenerate the schema for each hub version along with the conversions to and from the actual ARM API versions. Hand coding all the required conversions doesn't scale across all the different Azure sevices, especially with the ongoing rate of change. 
+
+**Allow for hand coded conversions:** While we expect to use code generation to handle the vast majority of needed conversions, we anticipate that some breaking API changes will require some of the conversion to be hand coded. We need to make it simple for these conversions to be introduced while still automatically generating the majority of the conversion.
+
+**No modification of generated files.** Manual modification of generated files is a known antipattern that greatly increases the complexity and burden of updates. If some files have been manually changed, every difference showing after code generation needs to be manually reviewed before being committed. This is tedious and error prone because the vast majority of auto generated changes will be perfectly fine. 
+
+**Compliance with Kubernetes versioning.** To quote [Kubebuilder's documentation](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html):
+
+> In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information. Thus, any change we make to our API must be compatible with whatever we supported in v1, and also need to make sure anything we add in v2 is supported in v1.
+
+## Non-Goals
+
+**Coverage of every case by code generation:** While it's likely that very high coverage will be achievable with code generation, we don't believe that it will be practical to handle every possible situation automatically. It's therefore necessary for the solution to have some form of extensibility allowing for the injection of hand written code.
+
+## Proposed Solution
+
+In summary (details below):
+
+* Base the storage version on the latest available API version for each resource type
+* Augment the storage version with a name/value property bag for storing additional information
+* Generate the required `ConvertTo()` and `ConvertFrom()` methods automatically
+* Allow configuration to support common changes from version to version
```

Comment:
This is supposed to be a very high level summary of the approach, with sections below going into details. I'll add more introductory text and also make the working clearer.

Reply:
I've completely reworked the lead-in to avoid digging too far into details too quickly. I've also pull out the examples into a pair of parallel case studies to allow comparison of the two major approaches.